### PR TITLE
[CL-314] improve `TableDataSource` filtering 

### PR DESCRIPTION
**Pull request by willmartian**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/CL-314

## 📔 Objective

Currently, devs often filter data both inside and outside of a `TableDataSource` it they want to extend the default filtering functionality. 

```ts
allUsers: OrganizationUserView[];

constructor() {
  // use filter implementation for free text search
  this.searchControl.valueChanges.subscribe(v => this.dataSource.filter = v));
}

filterByStatus(status: OrganizationUserStatusType | null) {
  if (status == null) {
    this.dataSource.data = this.allUsers;
    return;
  }

  this.dataSource.data = this.allUsers.filter(u => u.status == status);
}
```

This PR makes a few small changes to make this 1.) less confusing and 2.) more extensible when conditionally applying many arbitrary filters:

- combine `filter` and `filterPredicate` into a single field to allow for easier overriding
- expose the existing default filter functionality in a static method `TableDataSource.simpleStringFilter`

```ts
allUsers: OrganizationUserView[];

ngOnInit() {
  combineLatest([
    this.searchControl.valueChanges,
    this.filterByStatus$,
    // ...
  ]).subscribe(([
    searchValue,
    filterByStatus
  ]) => {
    this.dataSource.filter = (data) => 
      TableDataSource.simpleStringFilter(searchValue)(data) 
      && filterByStatus ? data.status === this.status : true
  });
}
```

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/components/src/table/table-data-source.ts [hunk 1]
```diff
@@ -207,36 +214,39 @@ export class TableDataSource<T> extends DataSource<T> {
...
-        return currentTerm + (data as unknown as Record<string, any>)[key] + "◬";
-      }, "")
-      .toLowerCase();
-
-    // Transform the filter by converting it to lowercase and removing whitespace.
-    const transformedFilter = filter.trim().toLowerCase();
-
-    return dataStr.indexOf(transformedFilter) != -1;
-  }
+  static simpleStringFilter = <T>(filter: string): FilterFn<T> => {
```
Comment (Hinton) [inline libs/components/src/table/table-data-source.ts:227]: **question**: Should we make this readonly so people don't do `TableDataSource.simpleStringFilter = "hack"`

Comment (willmartian) [reply libs/components/src/table/table-data-source.ts]: Yes, added in [ddb2afb](https://github.com/bitwarden/clients/pull/9708/commits/ddb2afb9b97b581bf85c1d6ec4a9bab0f12dc2c9).

**Comment by vleague2** [comment]
This could go in a follow-up ticket, but should we update the Storybook docs with how to implement custom filtering?

**Review by Hinton — DISMISSED** [review]

**Review by eliykat — APPROVED** [review]
Excited to start using this 🎉

**Review by Hinton — APPROVED** [review]

**Pull request merged**